lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,253 @@
+    """
+
+    def __init__(self, weight, max_score, locator):
+        """
+        Creates a BlockRecord.
+
+        Params:
+            weight (float)
+            max_score (float)
+            locator (BlockUsageLocator or string)
+        """
+        self.weight = weight
+        self.max_score = max_score
+        # pylint: disable=protected-access
+        self.block_key = locator._to_string() if isinstance(locator, BlockUsageLocator) else locator
+
+    def to_dict(self):
+        """
+        Serialize this object to a dict object.
+        """
+        return {
+            'weight': self.weight,
+            'max_score': self.max_score,
+            'block_key': self.block_key,
+        }
+
+    @property
+    def locator(self):
+        """
+        Convenience property allowing a BlockUsageLocator to be returned from this object's serialized data.
+        """
+        return BlockUsageLocator._from_string(self.block_key)  # pylint: disable=protected-access
+
+
+class PersistentSubsectionGrade(TimeStampedModel):
+    """
+    A django model tracking persistent grades at the subsection level.
+
+    TODO: here are the other query patterns listed in the ticket that are not currently used. Should any of these
+    indices be built?
+        user_id, course_id, content_type, is_valid
+        course_id, edit-timestamp
+        edit-timestamp
+    """
+
+    class Meta(object):
+        index_together = [
+            ('user_id', 'usage_key')
+        ]
+
+        unique_together = (('user_id', 'usage_key'))
+
+    id = UnsignedBigIntAutoField(primary_key=True)  # pylint: disable=invalid-name
+    subtree_edited_date = models.DateTimeField('last content edit timestamp')
+    user_id = models.CharField(max_length=255)
+    earned_all = models.IntegerField()
+    possible_all = models.IntegerField()
+    earned_graded = models.IntegerField()
+    possible_graded = models.IntegerField()
+
+    course_id = CourseKeyField(max_length=255)
+    usage_key = UsageKeyField(max_length=255)
+    course_version = models.CharField('guid of latest course version', max_length=255)
+
+    visible_blocks = models.ForeignKey(VisibleBlocks)
+
+    def __unicode__(self):
+        """
+        Returns a string representation of this model.
+        """
+        return "PersistentSubsectionGrade for user {} in subsection {}".format(self.user_id, self.usage_key)
+
+    @classmethod
+    def save_grade(cls, **kwargs):
+        """
+        Wrapper for create_grade or update_grade, depending on which applies.
+        Takes the same arguments as both of thsoe methods.
+        """
+        try:
+            cls.update(**kwargs)
+        except cls.DoesNotExist:
+            cls.create(**kwargs)
+
+    @classmethod
+    def create(
+        cls,
+        user_id,
+        usage_key,
+        course_version,
+        subtree_edited_date,
+        earned_all,
+        possible_all,
+        earned_graded,
+        possible_graded,
+        visible_blocks,
+    ):
```

Comment:
This looks like the kind of thing that you'd write a custom manager for: https://docs.djangoproject.com/en/1.8/topics/db/managers/